File: packages/nitrowind/ios/HybridNativePlatform.swift
import Foundation
import UIKit

#if canImport(NitroModules)
import NitroModules
#endif

/**
 * iOS implementation of the `NativePlatform` HybridObject. Reads appearance,
 * dimensions, safe-area insets, orientation, font scale and RTL from UIKit and
 * pushes every change into the C++ engine via `NitrowindBridge`, while also
 * notifying JS listeners.
 */
final class HybridNativePlatform: HybridNativePlatformSpec {
  private var listeners:
    [(_ dependencies: [StyleDependency], _ runtime: RuntimeSnapshot, _ source: RuntimeChangeSource) -> Void] = []
  private var extraThemes: [String] = []
  private var currentThemeName: String = "light"
  private var overrideColorScheme: Int? = nil
  private var adaptiveThemeFollowsColorScheme: Bool = true
  private var lastSnapshot: RuntimeSnapshot? = nil

  override init() {
    super.init()
    NotificationCenter.default.addObserver(
      self, selector: #selector(onEnvironmentChange),
      name: UIDevice.orientationDidChangeNotification, object: nil)
    NotificationCenter.default.addObserver(
      self, selector: #selector(onEnvironmentChange),
      name: UIApplication.didBecomeActiveNotification, object: nil)
    NotificationCenter.default.addObserver(
      self, selector: #selector(onEnvironmentChange),
      name: UIContentSizeCategory.didChangeNotification, object: nil)
    // Dark/light appearance changes. React Native posts this both when the
    // system appearance flips *and* when JS calls `Appearance.setColorScheme`
    // (the theme toggle), because both mutate the window's trait collection.
    NotificationCenter.default.addObserver(
      self, selector: #selector(onEnvironmentChange),
      name: NSNotification.Name("RCTUserInterfaceStyleDidChangeNotification"), object: nil)
  }

  deinit {
    NotificationCenter.default.removeObserver(self)
  }

  // MARK: - Reads

  private func keyWindow() -> UIWindow? {
    return UIApplication.shared.connectedScenes
      .compactMap { ($0 as? UIWindowScene)?.keyWindow }
      .first
  }

  private func colorSchemeRaw() -> Int {
    if let overrideColorScheme { return overrideColorScheme }
    // Read from the key window so we observe `Appearance.setColorScheme`
    // overrides too — `UITraitCollection.current` only reflects the system
    // value outside of a view's layout pass.
    let style = keyWindow()?.traitCollection.userInterfaceStyle
      ?? UITraitCollection.current.userInterfaceStyle
    return style == .dark ? 1 : 0
  }

  private func syncAdaptiveThemeName() {
    if adaptiveThemeFollowsColorScheme {
      currentThemeName = colorSchemeRaw() == 1 ? "dark" : "light"
    }
  }

  private func screenSize() -> (Double, Double) {
    let bounds = keyWindow()?.bounds ?? UIScreen.main.bounds
    return (Double(bounds.width), Double(bounds.height))
  }

  private func safeInsets() -> (Double, Double, Double, Double) {
    let insets = keyWindow()?.safeAreaInsets ?? .zero
    return (Double(insets.top), Double(insets.right), Double(insets.bottom), Double(insets.left))
  }

  private func orientationRaw() -> Int {
    let (w, h) = screenSize()
    return w >= h ? 1 : 0
  }

  private func isRTLValue() -> Bool {
    return UIView.userInterfaceLayoutDirection(for: .unspecified) == .rightToLeft
  }

  private func makeSnapshot() -> RuntimeSnapshot {
    syncAdaptiveThemeName()
    let (w, h) = screenSize()
    let (top, right, bottom, left) = safeInsets()
    let scale = Double(UIScreen.main.scale)
    return RuntimeSnapshot(
      colorScheme: colorSchemeRaw() == 1 ? .dark : .light,
      hasAdaptiveThemes: true,
      currentThemeName: currentThemeName,
      screen: Dimensions(width: w, height: h),
      insets: Insets(top: top, right: right, bottom: bottom, left: left),
      orientation: orientationRaw() == 1 ? .landscape : .portrait,
      pixelRatio: scale,
      fontScale: 1.0,
      rtl: isRTLValue(),
      rem: 16.0,
      hairlineWidth: 1.0 / scale)
  }

  private func pushToEngine() {
    syncAdaptiveThemeName()
    let (w, h) = screenSize()
    let (top, right, bottom, left) = safeInsets()
    let scale = Double(UIScreen.main.scale)
    NitrowindBridge.pushRuntimeState(
      withColorScheme: colorSchemeRaw(),
      themeName: currentThemeName,
      width: w, height: h,
      insetTop: top, insetRight: right, insetBottom: bottom, insetLeft: left,
      orientation: orientationRaw(),
      pixelRatio: scale,
      fontScale: 1.0,
      rtl: isRTLValue(),
      rem: 16.0,
      hairlineWidth: 1.0 / scale)
  }

  private func remeasureContainersSoon() {
    DispatchQueue.main.async {
      NitrowindBridge.remeasureContainers()
    }
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) {
      NitrowindBridge.remeasureContainers()
    }
  }

  private func dependencies(from previous: RuntimeSnapshot?, to next: RuntimeSnapshot) -> [StyleDependency] {
    guard let previous else { return [] }
    var deps: [StyleDependency] = []
    if previous.currentThemeName != next.currentThemeName { deps.append(.theme) }
    if previous.colorScheme != next.colorScheme { deps.append(.colorscheme) }
    if previous.screen.width != next.screen.width || previous.screen.height != next.screen.height { deps.append(.dimensions) }
    if previous.insets.top != next.insets.top || previous.insets.right != next.insets.right || previous.insets.bottom != next.insets.bottom || previous.insets.left != next.insets.left { deps.append(.insets) }
    if previous.orientation != next.orientation { deps.append(.orientation) }
    if previous.fontScale != next.fontScale { deps.append(.fontscale) }
    if previous.rtl != next.rtl { deps.append(.rtl) }
    return deps
  }

  @objc private func onEnvironmentChange() {
    let previous = lastSnapshot
    pushToEngine()
    remeasureContainersSoon()
    let snapshot = makeSnapshot()
    let deps = dependencies(from: previous, to: snapshot)
    lastSnapshot = snapshot
    if deps.isEmpty { return }
    for listener in listeners {
      listener(deps, snapshot, .system)
    }
  }

  private func emitUserThemeChange(_ deps: [StyleDependency]) {
    pushToEngine()
    let snapshot = makeSnapshot()
    lastSnapshot = snapshot
    if deps.isEmpty { return }
    for listener in listeners {
      listener(deps, snapshot, .user)
    }
  }

  // MARK: - Spec

  func getThemeConfig() -> ThemeConfig {
    syncAdaptiveThemeName()
    var themes = ["light", "dark"]
    themes.append(contentsOf: extraThemes)
    return ThemeConfig(themes: themes, currentTheme: currentThemeName, hasAdaptiveThemes: true)
  }

  func setTheme(theme: String) {
    let previousTheme = currentThemeName
    adaptiveThemeFollowsColorScheme = false
    currentThemeName = theme
    NitrowindBridge.setTheme(currentThemeName)
    var deps: [StyleDependency] = []
    if previousTheme != currentThemeName { deps.append(.theme) }
    emitUserThemeChange(deps)
  }

  func setColorScheme(scheme: ColorSchemeMode) {
    let previousColorScheme = colorSchemeRaw()
    let previousTheme = currentThemeName
    adaptiveThemeFollowsColorScheme = true
    if scheme == .system {
      overrideColorScheme = nil
    } else if scheme == .dark {
      overrideColorScheme = 1
    } else {
      overrideColorScheme = 0
    }
    syncAdaptiveThemeName()
    var deps: [StyleDependency] = []
    if previousColorScheme != colorSchemeRaw() { deps.append(.colorscheme) }
    if previousTheme != currentThemeName { deps.append(.theme) }
    emitUserThemeChange(deps)
  }

  func registerExtraThemes(themes: [String]) {
    extraThemes = themes
  }

  func getCurrent() -> RuntimeSnapshot {
    pushToEngine()
    let snapshot = makeSnapshot()
    lastSnapshot = snapshot
    return snapshot
  }

  func getColorScheme() -> ColorScheme { colorSchemeRaw() == 1 ? .dark : .light }

  func getDimensions() -> Dimensions {
    let (w, h) = screenSize()
    return Dimensions(width: w, height: h)
  }

  func getInsets() -> Insets {
    let (top, right, bottom, left) = safeInsets()
    return Insets(top: top, right: right, bottom: bottom, left: left)
  }

  func getOrientation() -> Orientation { orientationRaw() == 1 ? .landscape : .portrait }

  func getFontScale() -> Double { 1.0 }

  func getPixelRatio() -> Double { Double(UIScreen.main.scale) }

  func getIsRTL() -> Bool { isRTLValue() }

  func addRuntimeChangeListener(
    listener: @escaping (_ dependencies: [StyleDependency], _ runtime: RuntimeSnapshot, _ source: RuntimeChangeSource) -> Void
  ) {
    listeners.append(listener)
    // Prime the engine + listener with the current values.
    pushToEngine()
    let snapshot = makeSnapshot()
    lastSnapshot = snapshot
    listener([], snapshot, .system)
  }
}

File: packages/nitrowind/android/src/main/java/com/nitrowind/HybridNativePlatform.kt
package com.margelo.nitro.nitrowind

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.res.Configuration
import android.content.res.Resources
import android.os.Handler
import android.os.Looper
import android.view.View
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.facebook.proguard.annotations.DoNotStrip
import com.nitrowind.NitrowindContextHolder
import com.nitrowind.NitrowindNative
import java.util.Locale

/**
 * Android implementation of the `NativePlatform` HybridObject. Reads appearance,
 * dimensions, orientation, font scale and RTL from the system configuration and
 * forwards every change into the C++ engine (via [NitrowindNative.setRuntimeState])
 * while notifying JS listeners.
 */
@DoNotStrip
class HybridNativePlatform : HybridNativePlatformSpec() {
  override val memorySize: Long
    get() = 0L

  private val listeners =
    mutableListOf<(Array<StyleDependency>, RuntimeSnapshot, RuntimeChangeSource) -> Unit>()
  private var extraThemes: Array<String> = emptyArray()
  private var currentThemeName: String = "light"
  private var overrideColorScheme: Int? = null
  private var adaptiveThemeFollowsColorScheme: Boolean = true
  private var receiver: BroadcastReceiver? = null
  private var insetsListenerAttached: Boolean = false
  private var lastSnapshot: RuntimeSnapshot? = null
  private val mainHandler = Handler(Looper.getMainLooper())

  // MARK: - Reads

  private fun config(): Configuration = Resources.getSystem().configuration

  private fun colorSchemeRaw(): Int {
    overrideColorScheme?.let { return it }
    val night = config().uiMode and Configuration.UI_MODE_NIGHT_MASK
    return if (night == Configuration.UI_MODE_NIGHT_YES) 1 else 0
  }

  private fun syncAdaptiveThemeName() {
    if (adaptiveThemeFollowsColorScheme) {
      currentThemeName = if (colorSchemeRaw() == 1) "dark" else "light"
    }
  }

  private fun density(): Double = Resources.getSystem().displayMetrics.density.toDouble()

  private fun screenSize(): Pair<Double, Double> {
    val metrics = Resources.getSystem().displayMetrics
    val d = density().coerceAtLeast(1e-6)
    return Pair(metrics.widthPixels / d, metrics.heightPixels / d)
  }

  /** Live safe-area insets (system bars + display cutout) in dp. */
  private fun readInsets(): Insets {
    val decor: View = NitrowindContextHolder.currentActivity?.window?.decorView
      ?: return Insets(top = 0.0, right = 0.0, bottom = 0.0, left = 0.0)
    val root = ViewCompat.getRootWindowInsets(decor)
      ?: return Insets(top = 0.0, right = 0.0, bottom = 0.0, left = 0.0)
    val bars = root.getInsets(
      WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout()
    )
    val d = density().coerceAtLeast(1e-6)
    return Insets(
      top = bars.top / d,
      right = bars.right / d,
      bottom = bars.bottom / d,
      left = bars.left / d
    )
  }

  private fun orientationRaw(): Int =
    if (config().orientation == Configuration.ORIENTATION_LANDSCAPE) 1 else 0

  private fun isRtlValue(): Boolean =
    config().layoutDirection == View.LAYOUT_DIRECTION_RTL ||
      java.text.Bidi(
        Locale.getDefault().displayName,
        java.text.Bidi.DIRECTION_DEFAULT_LEFT_TO_RIGHT
      ).isRightToLeft

  private fun makeSnapshot(): RuntimeSnapshot {
    syncAdaptiveThemeName()
    val (w, h) = screenSize()
    val scale = density()
    return RuntimeSnapshot(
      colorScheme = if (colorSchemeRaw() == 1) ColorScheme.DARK else ColorScheme.LIGHT,
      hasAdaptiveThemes = true,
      currentThemeName = currentThemeName,
      screen = Dimensions(w, h),
      insets = readInsets(),
      orientation = if (orientationRaw() == 1) Orientation.LANDSCAPE else Orientation.PORTRAIT,
      pixelRatio = scale,
      fontScale = config().fontScale.toDouble(),
      rtl = isRtlValue(),
      rem = 16.0,
      hairlineWidth = 1.0 / scale.coerceAtLeast(1e-6)
    )
  }

  private fun pushToEngine() {
    syncAdaptiveThemeName()
    val (w, h) = screenSize()
    val scale = density()
    val insets = readInsets()
    NitrowindNative.setRuntimeState(
      colorScheme = colorSchemeRaw(),
      themeName = currentThemeName,
      width = w,
      height = h,
      insetTop = insets.top,
      insetRight = insets.right,
      insetBottom = insets.bottom,
      insetLeft = insets.left,
      orientation = orientationRaw(),
      pixelRatio = scale,
      fontScale = config().fontScale.toDouble(),
      rtl = isRtlValue(),
      rem = 16.0,
      hairlineWidth = 1.0 / scale.coerceAtLeast(1e-6)
    )
  }

  private fun remeasureContainersSoon() {
    mainHandler.post { NitrowindNative.remeasureContainers() }
    mainHandler.postDelayed({ NitrowindNative.remeasureContainers() }, 50)
  }

  private fun dependencies(previous: RuntimeSnapshot?, next: RuntimeSnapshot): Array<StyleDependency> {
    if (previous == null) return emptyArray()
    return buildList {
      if (previous.currentThemeName != next.currentThemeName) add(StyleDependency.THEME)
      if (previous.colorScheme != next.colorScheme) add(StyleDependency.COLORSCHEME)
      if (previous.screen.width != next.screen.width || previous.screen.height != next.screen.height) add(StyleDependency.DIMENSIONS)
      if (previous.insets.top != next.insets.top || previous.insets.right != next.insets.right || previous.insets.bottom != next.insets.bottom || previous.insets.left != next.insets.left) add(StyleDependency.INSETS)
      if (previous.orientation != next.orientation) add(StyleDependency.ORIENTATION)
      if (previous.fontScale != next.fontScale) add(StyleDependency.FONTSCALE)
      if (previous.rtl != next.rtl) add(StyleDependency.RTL)
    }.toTypedArray()
  }

  private fun emitChange() {
    val previous = lastSnapshot
    pushToEngine()
    remeasureContainersSoon()
    val snapshot = makeSnapshot()
    val deps = dependencies(previous, snapshot)
    lastSnapshot = snapshot
    if (deps.isEmpty()) return
    listeners.toList().forEach { it(deps, snapshot, RuntimeChangeSource.SYSTEM) }
  }

  private fun emitUserThemeChange(deps: Array<StyleDependency>) {
    pushToEngine()
    val snapshot = makeSnapshot()
    lastSnapshot = snapshot
    if (deps.isEmpty()) return
    listeners.toList().forEach { it(deps, snapshot, RuntimeChangeSource.USER) }
  }

  private fun ensureReceiver() {
    if (receiver != null) return
    val context: Context = NitrowindContextHolder.appContext ?: return
    val r = object : BroadcastReceiver() {
      override fun onReceive(ctx: Context?, intent: Intent?) {
        emitChange()
      }
    }
    context.registerReceiver(r, IntentFilter(Intent.ACTION_CONFIGURATION_CHANGED))
    receiver = r
  }

  /** Recompute natively whenever the window insets change (keyboard, cutout, rotation). */
  private fun ensureInsetsListener() {
    if (insetsListenerAttached) return
    val decor = NitrowindContextHolder.currentActivity?.window?.decorView ?: return
    insetsListenerAttached = true
    mainHandler.post {
      ViewCompat.setOnApplyWindowInsetsListener(decor) { _, windowInsets ->
        emitChange()
        windowInsets
      }
      ViewCompat.requestApplyInsets(decor)
    }
  }

  // MARK: - Spec

  override fun getThemeConfig(): ThemeConfig {
    syncAdaptiveThemeName()
    val themes = (listOf("light", "dark") + extraThemes.toList()).toTypedArray()
    return ThemeConfig(themes, currentThemeName, true)
  }

  override fun setTheme(theme: String) {
    val previousTheme = currentThemeName
    adaptiveThemeFollowsColorScheme = false
    currentThemeName = theme
    val deps = buildList {
      if (previousTheme != currentThemeName) add(StyleDependency.THEME)
    }.toTypedArray()
    emitUserThemeChange(deps)
  }

  override fun setColorScheme(scheme: ColorSchemeMode) {
    val previousColorScheme = colorSchemeRaw()
    val previousTheme = currentThemeName
    adaptiveThemeFollowsColorScheme = true
    overrideColorScheme = when (scheme) {
      ColorSchemeMode.SYSTEM -> null
      ColorSchemeMode.DARK -> 1
      else -> 0
    }
    syncAdaptiveThemeName()
    val deps = buildList {
      if (previousColorScheme != colorSchemeRaw()) add(StyleDependency.COLORSCHEME)
      if (previousTheme != currentThemeName) add(StyleDependency.THEME)
    }.toTypedArray()
    emitUserThemeChange(deps)
  }

  override fun registerExtraThemes(themes: Array<String>) {
    extraThemes = themes
  }

  override fun getCurrent(): RuntimeSnapshot {
    ensureReceiver()
    ensureInsetsListener()
    pushToEngine()
    val snapshot = makeSnapshot()
    lastSnapshot = snapshot
    return snapshot
  }

  override fun getColorScheme(): ColorScheme =
    if (colorSchemeRaw() == 1) ColorScheme.DARK else ColorScheme.LIGHT

  override fun getDimensions(): Dimensions {
    val (w, h) = screenSize()
    return Dimensions(w, h)
  }

  override fun getInsets(): Insets = readInsets()

  override fun getOrientation(): Orientation =
    if (orientationRaw() == 1) Orientation.LANDSCAPE else Orientation.PORTRAIT

  override fun getFontScale(): Double = config().fontScale.toDouble()

  override fun getPixelRatio(): Double = density()

  override fun getIsRTL(): Boolean = isRtlValue()

  override fun addRuntimeChangeListener(
    listener: (dependencies: Array<StyleDependency>, runtime: RuntimeSnapshot, source: RuntimeChangeSource) -> Unit
  ) {
    listeners.add(listener)
    ensureReceiver()
    ensureInsetsListener()
    // Prime the engine + listener with current values.
    pushToEngine()
    val snapshot = makeSnapshot()
    lastSnapshot = snapshot
    listener(emptyArray(), snapshot, RuntimeChangeSource.SYSTEM)
  }
}
